File: api/src/services/editor/file_filter.py
"""
File filtering for workspace operations.

Provides consistent filtering of system/metadata files across:
- File uploads (frontend and backend)
- File listings
- File writes

This module is the single source of truth for what files should be
excluded from the workspace.
"""

from pathlib import Path

# Directories to exclude from workspace operations
EXCLUDED_DIRECTORIES = frozenset({
    '.git',
    '__pycache__',
    '.vscode',
    '.idea',
    'node_modules',
    '.venv',
    'venv',
    'env',
    '.pytest_cache',
    '.mypy_cache',
    '.ruff_cache',
    'htmlcov',
    '.tox',
    '.nox',
    '.eggs',
    '*.egg-info',
    '.ipynb_checkpoints',
})

# Files to exclude from workspace operations
EXCLUDED_FILES = frozenset({
    '.DS_Store',
    'Thumbs.db',
    'desktop.ini',
    'bifrost.pyi',
    '.coverage',
    '.python-version',
    '.env',
    '.env.local',
})

# File extensions to exclude
EXCLUDED_EXTENSIONS = frozenset({'.pyc', '.pyo', '.pyd', '.so', '.dylib'})

# Prefixes that indicate hidden/metadata files
EXCLUDED_PREFIXES = ('._',)  # AppleDouble metadata files


def is_excluded_path(path: str | Path) -> bool:
    """
    Check if a path should be excluded from workspace operations.

    Works with both string paths and Path objects.
    Checks each component of the path against exclusion rules.

    Args:
        path: File path to check (relative or absolute)

    Returns:
        True if the path should be excluded, False otherwise
    """
    if isinstance(path, str):
        path = Path(path)

    # Check each component of the path
    for part in path.parts:
        # Skip empty parts and root
        if not part or part == '/':
            continue

        # Check hidden prefixes (AppleDouble files)
        for prefix in EXCLUDED_PREFIXES:
            if part.startswith(prefix):
                return True

        # Check exact matches (files and directories)
        if part in EXCLUDED_FILES or part in EXCLUDED_DIRECTORIES:
            return True

    # Check file extension of the final component
    if path.suffix.lower() in EXCLUDED_EXTENSIONS:
        return True

    return False


def is_allowed_path(path: str | Path) -> bool:
    """
    Check if a path is allowed for workspace operations.

    Inverse of is_excluded_path() for convenience.

    Args:
        path: File path to check

    Returns:
        True if the path is allowed, False if it should be excluded
    """
    return not is_excluded_path(path)


# Export constants for use in other modules (e.g., frontend type generation)
def get_exclusion_rules() -> dict:
    """
    Get all exclusion rules as a dictionary.

    Useful for serializing to JSON for frontend consumption.

    Returns:
        Dictionary with all exclusion rules
    """
    return {
        'directories': sorted(EXCLUDED_DIRECTORIES),
        'files': sorted(EXCLUDED_FILES),
        'extensions': sorted(EXCLUDED_EXTENSIONS),
        'prefixes': list(EXCLUDED_PREFIXES),
    }

File: client/src/lib/file-filter.ts
/**
 * File filtering for workspace operations.
 *
 * Provides consistent filtering of system/metadata files.
 * This mirrors the backend file_filter.py - keep them in sync!
 */

// Directories to exclude from workspace operations
const EXCLUDED_DIRECTORIES = new Set([
	".git",
	"__pycache__",
	".vscode",
	".idea",
	"node_modules",
	".venv",
	"venv",
	"env",
	".pytest_cache",
	".mypy_cache",
	".ruff_cache",
	"htmlcov",
	".tox",
	".nox",
	".eggs",
	".ipynb_checkpoints",
]);

// Files to exclude from workspace operations
const EXCLUDED_FILES = new Set([
	".DS_Store",
	"Thumbs.db",
	"desktop.ini",
	"bifrost.pyi",
	".coverage",
	".python-version",
	".env",
	".env.local",
]);

// File extensions to exclude
const EXCLUDED_EXTENSIONS = new Set([".pyc", ".pyo", ".pyd", ".so", ".dylib"]);

// Prefixes that indicate hidden/metadata files
const EXCLUDED_PREFIXES = ["._"]; // AppleDouble metadata files

/**
 * Check if a path should be excluded from workspace operations.
 *
 * Works with file paths (e.g., "folder/subfolder/file.txt").
 * Checks each component of the path against exclusion rules.
 *
 * @param path - File path to check (relative)
 * @returns True if the path should be excluded, False otherwise
 */
export function isExcludedPath(path: string): boolean {
	// Split path into components
	const parts = path.split("/").filter((p) => p.length > 0);

	for (const part of parts) {
		// Check hidden prefixes (AppleDouble files)
		for (const prefix of EXCLUDED_PREFIXES) {
			if (part.startsWith(prefix)) {
				return true;
			}
		}

		// Check exact matches (files and directories)
		if (EXCLUDED_FILES.has(part) || EXCLUDED_DIRECTORIES.has(part)) {
			return true;
		}

		// Check for egg-info directories (pattern match)
		if (part.endsWith(".egg-info")) {
			return true;
		}
	}

	// Check file extension of the final component
	const lastPart = parts[parts.length - 1];
	if (lastPart) {
		const dotIndex = lastPart.lastIndexOf(".");
		if (dotIndex > 0) {
			const extension = lastPart.substring(dotIndex).toLowerCase();
			if (EXCLUDED_EXTENSIONS.has(extension)) {
				return true;
			}
		}
	}

	return false;
}

/**
 * Check if a path is allowed for workspace operations.
 *
 * Inverse of isExcludedPath() for convenience.
 *
 * @param path - File path to check
 * @returns True if the path is allowed, False if it should be excluded
 */
export function isAllowedPath(path: string): boolean {
	return !isExcludedPath(path);
}

/**
 * Check if a path is a system-generated .bifrost/ file (read-only).
 *
 * @param path - File path to check (relative)
 * @returns True if the path is under .bifrost/ and should be read-only
 */
export function isBifrostSystemFile(path: string): boolean {
	return path === ".bifrost" || path.startsWith(".bifrost/");
}
